### PR TITLE
Chained Auth

### PR DESCRIPTION
**Pull request by jrouzierinverse**
## Description

Allows to restrict the values for for Access Levels, Access Durations, and User Roles depending on admin role.
You can chain an internal source to an external source
## Impacts
- Admin Gui
  - Can you now restrict an 'Admin Roles' to only accept certain values for Access Levels, Access Durations, and User Roles.
  - The new chained auth source (Create/Read/Update/Delete)
  - The chained source has two fields authentication_source and chained_authentication_source
  - authentication_source can only be any internal source except chained sources
  - chained_authentication_source can only be any external sources
- Captival Portal
  - When users are authenticate against a chained auth source they are taken to the chained_authentication_source signup page.
  - The user is properly registered after completing completing the chained_authentication_source requirements
## Delete branch after merge

NO (Keeps your dirty paws off my repo)

### CODE REVIEW

**Comment by julsemaan** [comment]
Is there a reason why you mixed the chained auth with what you did in #271 ?

**Comment by julsemaan** [comment]
it doesn't work at all. authenticating with either the internal or external source grants access to the network.

**Comment by julsemaan** [comment]
configuring the sources in the config files directly and entering invalid sources (like putting local in the chained source) checks up. The check for the internal vs external sources should be done in the pfcmd checkup to prevent that

**Comment by julsemaan** [comment]
This applies to #271 and this one but when selecting which access duration the user can set, it's a field where you need to input the durations manually in a csv form. This should be a multiple select like everywhere else in the admin ui

File: lib/pf/authentication.pm [hunk 1]
```diff
@@ -250,7 +250,10 @@ sub _guest_modes_from_sources {
     my ($sources) = @_;
     $sources ||= [];
     my %is_in = map {$_ => undef } @$sources;
-    return join(',', map { lc($_->type)} grep { exists $is_in{$_->id} && $_->class eq 'external'} @authentication_sources);
+    my @modes = map { lc($_->type)} grep { exists $is_in{$_->id} && $_->class eq 'external'} @authentication_sources;
```
Comment (dwlfrth) [inline lib/pf/authentication.pm]: Just a little comment on what we want to achieve here will save some other a headache trying to figure it out

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm [hunk 2]
```diff
@@ -206,6 +207,60 @@ sub postAuthentication : Private {
...
+    SponsorEmail => 'sponsored_guest_allowed',
+);
+
+=head2 isGuestSigned
+
+Checks to see if the source is a signup source
+
+=cut
+
+
```
Comment (dwlfrth) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm]: Remove some blank lines here

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm [hunk 3]
```diff
@@ -206,6 +207,60 @@ sub postAuthentication : Private {
     $c->forward('checkIfProvisionIsNeeded');
 }
 
+=head2 checkIfChainedAuth
+
+Checked to see if source that was authenticated with is chained
+
+=cut
+
+sub checkIfChainedAuth : Private {
```
Comment (dwlfrth) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm:216]: Having one or two logging statement here would not kill anyone ;)
That's actually the only way the software can talk to us.

File: html/pfappserver/lib/pfappserver/Base/Form/Authentication/Action.pm [hunk 4]
```diff
@@ -154,16 +156,24 @@ Populate the select field for the roles template action.
...
 
-    my @roles;
+=head2 _get_allowed_options
 
-    # Build a list of existing roles
-    my ($status, $result) = $self->form->ctx->model('Roles')->list();
-    if (is_success($status)) {
-        @roles = map { $_->{name} => $_->{name} } @$result;
-    }
+TODO: documention
```
Comment (dwlfrth) [inline html/pfappserver/lib/pfappserver/Base/Form/Authentication/Action.pm]: Guess who's back !

File: html/pfappserver/lib/pfappserver/Form/Config/AdminRoles.pm [hunk 5]
```diff
@@ -15,6 +15,9 @@ extends 'pfappserver::Base::Form';
 with 'pfappserver::Base::Form::Role::Help';
 
 use pf::admin_roles;
+use pf::log;
```
Comment (dwlfrth) [inline html/pfappserver/lib/pfappserver/Form/Config/AdminRoles.pm:18]: in a Form ?

File: html/pfappserver/lib/pfappserver/Form/Config/Authentication/Source/Chained.pm [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+
+=head1 AUTHOR
+
+Inverse inc. <info@inverse.ca>
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2013 Inverse inc.
```
Comment (dwlfrth) [inline html/pfappserver/lib/pfappserver/Form/Config/Authentication/Source/Chained.pm]: Welcome in 2014! Almost 2015 ! ;)

File: html/pfappserver/lib/pfappserver/Form/Config/AdminRoles.pm (hunk 5 above)
Comment (jrouzierinverse) [inline html/pfappserver/lib/pfappserver/Form/Config/AdminRoles.pm:18]: I must have used for debugging should it be removed?

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm (hunk 3 above)
Comment (jrouzierinverse) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Authenticate.pm:216]: checkIfChainedAuth will be logged automatically in the log/httpd.portal.catalyst

**Comment by jrouzierinverse** [comment]
I am done with my changes.
Let me know if there is anything else needed.

File: html/pfappserver/lib/pfappserver/Form/Config/Authentication/Source/Chained.pm [hunk 7]
```diff
@@ -0,0 +1,108 @@
+package pfappserver::Form::Config::Authentication::Source::Chained;
+
+=head1 NAME
+
+pfappserver::Form::Config::Authentication::Source::Chained add documentation
```
Comment (julsemaan) [inline html/pfappserver/lib/pfappserver/Form/Config/Authentication/Source/Chained.pm]: can you add documentation or remove the comment

File: lib/pf/Authentication/Source/ChainedSource.pm [hunk 8]
```diff
@@ -0,0 +1,162 @@
+package pf::Authentication::Source::ChainedSource;
+=head1 NAME
+
+pf::Authentication::Source::ChainedSource add documentation
```
Comment (julsemaan) [inline lib/pf/Authentication/Source/ChainedSource.pm]: same here for the add documentation

File: lib/pf/admin_roles.pm [hunk 9]
```diff
@@ -182,6 +185,45 @@ our $cached_adminroles_config = pf::config::cached->new(
...
+    foreach my $role (@$roles) {
+        next unless exists $ADMIN_ROLES{$role};
+        #If no option is defined then all are allowed
+        return unless exists $ADMIN_ROLES{$role}{$option};
+
+        my $allowed_options = $ADMIN_ROLES{$role}{$option};
+        #If the allowed options is empty the all are allowed
+        return unless defined $allowed_options && length $allowed_options;
+
+        push @options, split /\s*,\s*/, $allowed_options;
```
Comment (julsemaan) [inline lib/pf/admin_roles.pm:227]: don't we have something that we can use instead of the split ?

File: lib/pf/admin_roles.pm [hunk 10]
```diff
@@ -182,6 +185,45 @@ our $cached_adminroles_config = pf::config::cached->new(
...
+
+=head2 admin_allowed_options_all
+
+Get all the allowed values for a given role
+
+=cut
+
+sub admin_allowed_options_all {
+    my ($roles, $option) = @_;
+    return uniq map {split /\s*,\s*/, ($ADMIN_ROLES{$_}{$option} || '')} grep { exists $ADMIN_ROLES{$_} && exists $ADMIN_ROLES{$_}{$option} }  @$roles;
```
Comment (julsemaan) [inline lib/pf/admin_roles.pm:241]: same for the split here

File: t/unittest/admin_roles.t [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+use Test::NoWarnings;
+
+use_ok("pf::admin_roles");
+
+ok(admin_can(["User Manager"],'USERS_CREATE'),"User Manager can create a user");
+
+ok(!admin_can(["User Manager"],'NODES_READ'),"User Manager cannot read a node");
+
+ok(admin_can(["ALL"],'USERS_CREATE'),"ALL can create a user");
+
```
Comment (julsemaan) [inline t/unittest/admin_roles.t:31]: would be useful to have tests for the new features here. like for how the roles are combined and make sure that the proper behavior happens.

File: lib/pf/admin_roles.pm (hunk 10 above)
Comment (jrouzierinverse) [inline lib/pf/admin_roles.pm:241]: Not at the moment.

**Comment by julsemaan** [comment]
James, can you rebase it so we merge it into devel or so I can generate a patch.

Thanks

**Comment by jrouzierinverse** [comment]
Done

James Rouzier
jrouzier@inverse.ca :: +1.514.755.3630  ::  http://www.inverse.ca
Inverse inc. :: Leaders behind SOGo (http://www.sogo.nu) and PacketFence (http://www.packetfence.org)

**Pull request merged**